File: android/xamarin.android/ProgrammingIdeas/Fragments/LoginDialog.cs
using Android.App;
using Android.OS;
using Android.Support.Design.Widget;
using Android.Views;
using Android.Widget;
using Helpers;
using Newtonsoft.Json;
using ProgrammingIdeas.Api;
using ProgrammingIdeas.Helpers;
using ProgrammingIdeas.Models;
using System;
using System.Threading.Tasks;
using AlertDialog = Android.Support.V7.App.AlertDialog;
using DialogFragment = Android.Support.V4.App.DialogFragment;

namespace ProgrammingIdeas.Fragments
{
    internal class LoginDialog : DialogFragment
    {
        private ProgressBar loadingCircle;

        public event EventHandler<string> OnUserLoggedIn;

        public override Dialog OnCreateDialog(Bundle savedInstanceState)
        {
            var view = LayoutInflater.From(App.CurrentActivity).Inflate(Resource.Layout.login_dialog, null, false);
            var emailTb = view.FindViewById<EditText>(Resource.Id.emailTb);
            var passwordTb = view.FindViewById<EditText>(Resource.Id.passwordTb);
            var loginBtn = view.FindViewById<Button>(Resource.Id.loginBtn);
            loadingCircle = view.FindViewById<ProgressBar>(Resource.Id.loadingCircle);

            loginBtn.Click += async delegate
            {
                using (var validator = new Validator())
                {
                    loadingCircle.Visibility = ViewStates.Visible;
                    loginBtn.Enabled = false;

                    validator.ValidateIsNotEmpty(emailTb, true);
                    validator.ValidateIsNotEmpty(passwordTb, true);

                    if (validator.PassedValidation)
                    {
                        var data = new LoginData { Email = emailTb.Text.Trim(), Password = passwordTb.Text.Trim() };
                        await LoginUser(data);
                    }

                    loginBtn.Enabled = true;
                    loadingCircle.Visibility = ViewStates.Gone;
                }
            };

            return new AlertDialog.Builder(Activity)
                .SetTitle("Login")
                .SetView(view)
                .Create();
        }

        private async Task LoginUser(LoginData data)
        {
            var response = await IdeaBagApi.Instance.LoginAsync(data);

            if (response.Payload != null)
            {
                Global.AuthUser(response.Payload);

                var expirationDate = DateTime.Now.AddMilliseconds(response.Payload.ExpiresInMillis);
                PreferenceManager.Instance.AddEntry("expiresIn", JsonConvert.SerializeObject(expirationDate));

                OnUserLoggedIn?.Invoke(this, response.Payload.Email);
                Dismiss();
            }
            else
                Snackbar.Make(loadingCircle, response.ErrorMessage, Snackbar.LengthLong).Show();
        }
    }
}

File: android/kotlin/app/src/main/java/com/alansa/ideabag2/dialogs/LoginDialog.kt
package com.alansa.ideabag2.dialogs

import android.support.v7.app.AlertDialog
import android.app.Dialog
import android.content.Context
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.view.LayoutInflater
import android.view.View
import android.widget.Toast
import com.alansa.ideabag2.R
import com.alansa.ideabag2.extensions.using
import com.alansa.ideabag2.utils.Validator
import com.google.firebase.auth.FirebaseAuth
import kotlinx.android.synthetic.main.dialog_login.view.*

class LoginDialog(private val ctx : Context, private val onLoginSuccess: () -> Unit) : DialogFragment() {
    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        var view = LayoutInflater.from(activity).inflate(R.layout.dialog_login, null)

        view.loginBtn.setOnClickListener {

            val validator = Validator()
            using(validator) {
                val email = view.emailTb.text.toString()
                val pwd = view.passwordTb.text.toString()
                validator.validateIsNotEmpty(view.emailTIL, email)
                validator.validateIsNotEmpty(view.passwordTIL, pwd)

                if (validator.passedValidation) {
                    view.loadingCircle.visibility = View.VISIBLE
                    view.loginBtn.isEnabled = false

                    FirebaseAuth
                            .getInstance()
                            .signInWithEmailAndPassword(email, pwd)
                            .addOnCompleteListener {
                                view.loadingCircle.visibility = View.GONE
                                view.loginBtn.isEnabled = true

                                Toast.makeText(ctx, if (it.isSuccessful) "Login successful." else it.exception?.message, Toast.LENGTH_LONG).show()

                                if (it.isSuccessful) {
                                    onLoginSuccess()
                                    dismiss()
                                }
                            }
                } else {
                    Toast.makeText(ctx, R.string.required, Toast.LENGTH_LONG).show()
                }
            }
        }

        return AlertDialog.Builder(ctx)
                .setTitle(R.string.login)
                .setView(view)
                .create()
    }
}